### ISSUE TITLE
Improve refund endpoint documentation

### ISSUE THREAD

**Issue by misterjoonas** [CheckoutFinland/psp-api#14]
Refund endpoint seems to have changed quite a lot. Is it really so that payments can be refunded with just the transaction ID? No merchant credentials nor hmac calculation needed? How can it be tested if not using the test credentials? What does the response look like? Is it now always `200 OK` before the async refunds are implemented? So we can only display "refund successful" always? When is the cancel callback url called?

**Comment by vesse** [comment]
Yes, we do need to improve the documentation. Probably that happens earliest next week. 

The note about async refunds is that we hope for merchants to be prepared for them. Currently the refunds are synchronous, and the response is as described in the OpenAPI specication, namely:

```json
{
  "status": "ok",
  "transactionId": "uuid",
  "provider": "osuuspankki"
}
```

and status codes are `20x` or `400`

**Comment by misterjoonas** [comment]
But is there a way to test refunding? I don't have a real transaction-id to use when testing the API. Or maybe any test payment can be test refunded? Does the stamp have to match the original stamp or generate a new stamp for the refund request?

**Comment by vesse** [comment]
You can refund the test payments normally, the response when you created the payment works.

Nordea test api does not support refunds, and S-pankki and Ålandsbanken don’t support API refunds at all, but for the other providera you can refund normally.

And the refund item stamp needs to be the original stamp for the item (applies to shop-in-shop only) as it is used to locate the original item

**Comment by jrasanen** [comment]
Yup! You can fully test refunds with the API, as Vesa mentioned. Merchant information is passed in http headers.

**Comment by jrasanen** [comment]
Here's a working test code:

```javascript
const crypto = require('crypto');
const SuperAgent = require('SuperAgent');

const ACCOUNT = '375917';
const SECRET = 'SAIPPUAKAUPPIAS';

const trxId = '70979c7a-b7df-11e8-96fe-77127a524514';

const infoHeaders = {
  'checkout-account': ACCOUNT,
  'checkout-algorithm': 'sha256',
  'checkout-method': 'POST',
  'checkout-nonce': process.hrtime().join(''),
  'checkout-timestamp': new Date().toISOString(),
  'checkout-transaction-id': trxId
};

const payload = {
  "amount": 100,
  "callbackUrls": {
    "success": "https://example.org",
    "cancel": "https://example.org"
  }
};


const hmacInfoPayload = Object.keys(infoHeaders)
    .sort()
    .map((key) => [ key, infoHeaders[key] ].join(':'))
    .concat(JSON.stringify(payload))
    .join("\n");

const hmacInfo = crypto
  .createHmac('sha256', SECRET)
  .update(hmacInfoPayload)
  .digest('hex');

SuperAgent
  .post('https://api.checkout.fi/payments/' + trxId + '/refund/')
  .set(Object.assign({}, infoHeaders, { signature: hmacInfo }))
  .send(payload)
  .then((res) => console.log(res.body))
  .catch((err) => console.error(err.response.body, err.response.header, err.status));
```

I created a trade for our test merchant, you should be able to run that code & refund it couple of times.

**Comment by misterjoonas** [comment]
What does it mean that S-Pankki and Ålandsbanken don't support refunds? Their clients cannot get their money back at all?
I have always used Nordea for testing payments. Is there a list somewhere of all providers which can be tested?

**Comment by jrasanen** [comment]
S-Pankki, Ålandsbanken has not implemented refund functionality at all :/ You have to create an email refund for those.

Nordea's API throws an error 'Palvelusopimus ei ole voimassa' for test refunds and they've documented that the refunds will work in production but test mode does not support it.

@misterjoonas we're going to update the documentation regarding on the testability. And, I'll keep this open to remind us of it!

Action points for us next week
- [ ] Email-refund documentation
- [ ] Mention payment providers which do not provide test mode for refunds

**Comment by misterjoonas** [comment]
Is this how the old refund API works too? I have only implemented the refund for a client, but I don't think it is mentioned anywhere that some banks do not approve refunds.

I thought checkout.fi exists for this purpose so that services can connect to one API no matter which bank their customers have.

**Comment by jrasanen** [comment]
@misterjoonas good feedback, thanks!

We've considered automatically fallback to email refunds, if a provider does not support those or they have technical difficulties, but haven't implemented this before getting more feedback on the matter. 

What do you think? Should we do that or add a flag to the refund API implementation, which allows the fallback option?

**Comment by misterjoonas** [comment]
I'm not sure what you are asking. I have been always under the impression that checkout.fi will handle all payments and refunds. Otherwise every service provider (client's of checkout.fi) would have to know about providers (banks) and their supported or non-supported services (APIs) and how to connect to them.

The old refund API (https://checkoutfinland.github.io/legacy-api/#refund-api) has a lot of different response status codes which can be displayed to user in case refund is not possible. The new one has two, 201 (success) and 422 (not supported). What should the service provider do in case of 422? Figure out which bank was used, how to reach them to make a refund, tell the customer to sit tight and wait? I don't think so.

I'd say at least three situations should be covered:
- Success: ("Your money has been refunded, yay!")
- Failed: ("For reason X, your money has not been refunded. Please contact customer service or the issuing bank.")
- Delayed: ("Your refund request has been forwarded to the issuing bank and should be processed soon. Please contact the bank in X days if the money has not been refunded to your account.")

My two cents.

**Comment by misterjoonas** [comment]
I tried to make a refund, but got a message _Service agreement not in force. (V2322A1N)_ for the 375917 account.

I posted my request to `https://api.checkout.fi/payments/aa1cf28e-b82e-11e8-ac39-df5cb846b49a/refund` with the following data:
```yaml
{
  "amount": 100,
  "callbackUrls": {
    "success": "https://local.project_name.fi/en/checkoutfi/test/refund/success",
    "cancel": "https://local.project_name.fi/en/checkoutfi/test/refund/cancel"
  }
}
```

**Comment by misterjoonas** [comment]
I tried to make a refund, but got a message _Service agreement not in force. (V2322A1N)_ for the 375917 account.

I posted my request to `https://api.checkout.fi/payments/aa1cf28e-b82e-11e8-ac39-df5cb846b49a/refund` with the following data:
```yaml
{
  "amount": 100,
  "callbackUrls": {
    "success": "https://local.project_name.fi/en/checkoutfi/test/refund/success",
    "cancel": "https://local.project_name.fi/en/checkoutfi/test/refund/cancel"
  }
}
```

**Comment by vesse** [comment]
That's because, as mentioned, Nordea test API does not support refunds.

The legacy API returned success also for Nordea refunds but only because it was hard-coded to the implementation to return success even when the call had failed. We didn't consider that a good option so the new one does not do that. On the upside, now you can test failed refunds with Nordea payments.

**Comment by misterjoonas** [comment]
Okay, so it seems that the response code is 400 in that case, which is not in the documentation nor at the specification at https://checkoutfinland.github.io/psp-api/checkout-psp-api.yaml
Is there a reason why the cancel callback url is not called in that case?

**Comment by vesse** [comment]
Yes. As I mentioned earlier:

> The note about async refunds is that we hope for merchants to be prepared for them. Currently the refunds are synchronous, and the response is as described in the OpenAPI specication, namely:
> 
> ```json
> {
>   "status": "ok",
>   "transactionId": "uuid",
>   "provider": "osuuspankki"
> }
> ```
> 
> and status codes are `20x` or `400`

(and the 422 for not supported). And the callback is not called for the same reason - since the API is synchronous the result is pretty obvious. The OpenAPI spec describes what we initially planned, but there were some early onboarders who needed synchronous refunds due to which the async implementation was postponed. I'm planning to update the refund documentation this week, and will change the YAML too to match what we currently have instead of what was initially planned.

**Comment by misterjoonas** [comment]
Finally I found some test credentials here http://docs.maksuturva.fi/fi/html/pages/4_2_1_verkkopankkien_testitunnukset.html for other banks than Nordea.
Made a payment `cc0231ca-ba72-11e8-9fef-1f94c0345cf4` with Aktia and tried to refund that, but now I got a weird message _Cannot parse refund response_ with code 400 on the response.

**Comment by vesse** [comment]
We are going to add the credentials to our documentation too for obvious reasons. I'll need to contact Aktia about the refunds in the test APIs. In your case the response was `<html><head><meta HTTP-EQUIV=\"REFRESH\" content=\"0; url=https://ebank.aktia.fi/error/aktia.msg?a=100013&b=W5@bTwr@zQ0AAAIhaP8AAACz\"></head><body></body></html>\n\n`

**Comment by vesse** [comment]
OK turns out Aktia does not work either in test mode, we would need to implement the same kind of hack as for OP since they have only one payment that can be refunded. So ATM this is where we stand with test account:

* Masterpass: refunds work but approving test payments is difficult (sometimes [these instructions](https://developer.mastercard.com/page/masterpass-sandbox-testing-guidelines) work in the web flow, mobile app flow cannot be tested)
* MobilePay: refunds work, payments can be accepted with your own MobilePay account, the charges are not made when using test merchant
* OP: refunds work, but as mentioned earlier they actually always refund the same payment in test mode
* Nordea: refunds do not work
* Handelsbanken: refunds work
* POP: refunds work
* Aktia: refunds do not work, would require same hack as for OP
* Säästöpankki: refunds work
* OmaSP: refunds work
* S-pankki: refunds not supported
* Ålandsbanken: refunds not supported
* Danske: refunds work, but approving payments requires Danske consumer credentials
* Credit cards: refunds work, use [these cards](https://dev.paymenthighway.io/#sandbox-credit-cards) when approving payments
* Collector bank: refunds work, use `010380-000P` for SSN when approving payment
* Mash: refunds work, use [genrated SSN](https://sc-rel.mash.com/My/Test/GenerateSsnForTesting?age=34&tps=651) when approving payment

Handelsbanken, POP, SP and OmaSP show their test account login information on the login pages so with them it's relatively fast to test payments, just copy-paste the values.

**Comment by misterjoonas** [comment]
Okay, now I'm getting forward with those banks.

It seems that I have to use a try-catch procedure and handle the request exception when making invalid requests to refund API. And the message is in the exception message, in a string together with some JSON data:
`Client error: 'POST https://api.checkout.fi/payments/35ce7edc-bb0f-11e8-8504-ff664d681859/refund' resulted in a '400 Bad Request' response:
{"status":"error","message":"Schema validation failed","meta":["Refund amount exceeds transaction amount"]}`

This is quite annoying to handle, instead of getting a proper response with a code (400 for example) and plain JSON in the body, as it is when I get a `201 "Created"` for a valid refund.
I guess user does not need to know anything else than the refund failed for one reason or another, but for debugging it would make sense to handle the different failures and log them properly. Now I can just pass the string to the logger and log a generic "refund failed" message. Which is better than nothing...

**Comment by vesse** [comment]
It is up to you how to handle the non-OK status codes, but the response is a valid HTTP response with correct `content-type` header, and if the library of your choice does not parse the body to a JSON object then maybe change the library. Nothing we can help with as for very obvious reasons we are not going to start returning non-OK replies with OK HTTP status.

**Comment by misterjoonas** [comment]
I was not using the `ignore_errors` you had in the example. I'm creating the request with a GuzzleHttp client, so I added a `RequestOptions::HTTP_ERRORS => FALSE` to the request configuration and now there is no exception on 4xx or 5xx response.

But it seems that the browser is not redirected to the success/cancel urls which are in the example, do I need to handle them too somehow? I tried to set `RequestOptions::ALLOW_REDIRECTS => TRUE` but no difference. It doesn't matter right now because I have a test form currently which will display error message for user, but I'm interested if the redirect should happen or not.
